python/paddle/distributed/passes/gradient_merge.py

```diff
@@ -0,0 +1,349 @@
+        persistable=False,
+        force_cpu=True)
+
+    with device_guard("cpu"):
+        # step_var = (step_var + 1) % k_step
+        layers.increment(x=step_var, value=1.0, in_place=True)
+        main_block.append_op(
+            type='elementwise_mod',
+            inputs={'X': step_var,
+                    'Y': k_step_var},
+            outputs={'Out': step_var},
+            attrs={'axis': -1,
+                   'use_mkldnn': False})
+
+        # cond_var = (step_var == 0)
+        main_block.append_op(
+            type='equal',
+            inputs={'X': step_var,
+                    'Y': zero_var},
+            outputs={'Out': cond_var})
+
+    return cond_var
+
+
+def _append_gradient_merge_backward_op(main_program,
+                                       startup_program,
+                                       params_grads: list,
+                                       cond_var_name: str):
+    main_block = main_program.global_block()
+    startup_block = startup_program.global_block()
+
+    #TODO(mapingshuo) support sparse embedding
+    # step1: remove grad.op's op_role_var
+    for param, grad in params_grads:
+        assert (
+            param.type != core.VarDesc.VarType.SELECTED_ROWS
+        ), "SELECTED_ROWS is not supported in GradientMergeOptimizer for now"
+
+        _remove_op_role_var(param, grad)
+
+    param_to_gradient_merge = {}
+    new_params_grads = []
+    # step2: create gradient_merge var and init with 0
+    # and update op_role_var
+    for param, grad in params_grads:
+        param_name = param.name
+        param_var = main_block.var(param_name)
+        assert (param_var is not None)
+        gradient_merge_var = main_block.create_var(
+            name=param_name + "@GRAD@GradientMerge",
+            shape=param_var.shape,
+            dtype=param_var.dtype,
+            persistable=True)
+        param_to_gradient_merge[param_name] = gradient_merge_var
+
+        startup_gradient_merge_var = startup_block.create_var(
+            name=param_name + "@GRAD@GradientMerge",
+            shape=param_var.shape,
+            dtype=param_var.dtype,
+            persistable=True)
+        startup_block.append_op(
+            type="fill_constant",
+            outputs={"Out": startup_gradient_merge_var},
+            attrs={
+                "shape": param_var.shape,
+                "dtype": param_var.dtype,
+                "value": float(0),
+            })
+
+        # grad_merge += grad
+        new_grad_op = main_block.append_op(
+            type="elementwise_add",
+            inputs={'X': grad,
+                    'Y': gradient_merge_var},
+            outputs={'Out': gradient_merge_var},
+            attrs={'axis': -1,
+                   'use_mkldnn': False})
+        _add_gm_op_role_var(new_grad_op, param, gradient_merge_var,
+                            cond_var_name)
+        new_params_grads.append([param, gradient_merge_var])
+    return new_params_grads, param_to_gradient_merge
```

Comment:
It is better to rename `new_params_grads` to `new_params_to_grads` as `param_to_gradient_merge` to explicitly    indicate a dict.

Reply:
done